pkg/storage/unified/resource/datastore.go

```diff
@@ -274,20 +420,221 @@ func ParseKey(key string) (DataKey, error) {
+			processLastResource()
+		}
+
+		// Update tracking variables for the current resource
+		currentResourceKey = resourceKey
+		lastDataKey = &dataKey
+	}
+
+	// Process the final resource
+	processLastResource()
+
+	// Convert namespace counts to ResourceStats
+	stats := make([]ResourceStats, 0, len(namespaceCounts))
+	for ns, names := range namespaceCounts {
+		// Count how many names actually exist (not deleted)
+		count := int64(0)
+		for _, exists := range names {
+			if exists {
+				count++
+			}
+		}
+
+		// Skip if count is below or equal to minimum
+		if count <= int64(minCount) {
+			continue
+		}
+
+		stats = append(stats, ResourceStats{
+			NamespacedResource: NamespacedResource{
+				Namespace: ns,
+				Group:     groupResource.Group,
+				Resource:  groupResource.Resource,
+			},
+			Count:           count,
+			ResourceVersion: namespaceVersions[ns],
+		})
+	}
+
+	return stats, nil
+}
+
+// getGroupResources returns all unique group/resource combinations in the data store.
+// It efficiently discovers these by using the key ordering and PrefixRangeEnd to jump
+// between different group/resource prefixes without iterating through all keys.
+// Results are cached to improve performance.
+func (d *dataStore) getGroupResources(ctx context.Context) ([]GroupResource, error) {
+	// Check cache first
+	if cached, found := d.cache.Get(groupResourcesCacheKey); found {
+		if cachedResults, ok := cached.([]GroupResource); ok {
+			return cachedResults, nil
+		}
+	}
+
+	// Cache miss or invalid data, compute the results
+	results := make([]GroupResource, 0)
+	seenGroupResources := make(map[string]bool) // "group/resource" -> seen
+
+	startKey := ""
+
+	for {
+		// List with limit 1 to get the next key
+		var foundKey string
+		var foundAny bool
+
+		for key, err := range d.kv.Keys(ctx, dataSection, ListOptions{
+			StartKey: startKey,
+			Limit:    1,
+			Sort:     SortOrderAsc,
+		}) {
+			if err != nil {
+				return nil, err
+			}
+			foundKey = key
+			foundAny = true
+			break // Only process the first (and only) key
+		}
+
+		// If no key found, we're done
+		if !foundAny {
+			break
+		}
+
+		// Parse the key to extract group and resource
+		dataKey, err := ParseKey(foundKey)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse key %s: %w", foundKey, err)
+		}
+
+		// Create the group/resource identifier
+		groupResourceKey := fmt.Sprintf("%s/%s", dataKey.Group, dataKey.Resource)
+
+		// Add to results if we haven't seen this group/resource combination before
+		if !seenGroupResources[groupResourceKey] {
+			seenGroupResources[groupResourceKey] = true
+			//nolint:staticcheck // SA4010: wrongly assumes that this result of append is never used
+			results = append(results, GroupResource{
+				Group:    dataKey.Group,
+				Resource: dataKey.Resource,
+			})
+		}
+
+		// Compute the next starting point by finding the end of this group/resource prefix
+		groupResourcePrefix := fmt.Sprintf("%s/%s/", dataKey.Group, dataKey.Resource)
```

Comment:
Trying to wrap my head around `PrefixRangeEnd()`. Since it increments the trailing byte, do we always need to include the trailing slash in the key? Do we consider the trailing slash part of the key?